### PR TITLE
fix: bundled wasm assets not found during build

### LINKED ISSUE #672: [🐛 Bug]: Unexpected error: ENOENT: no such file or directory
### next-on-pages environment related information

_No response_

### Description

Cloudflare build fails when using git integration on a minimal repository containing `create-next-app` and an additional file located at `src/app/og/route.tsx`.

### Reproduction

Minimal reproduction: https://github.com/morganfeeney/minimal-repro

### Pages Deployment Method

Pages CI (GitHub/GitLab integration)

### Pages Deployment ID

7bef422752a3c808130c2294019a891d

### Additional Information

2024-02-20T20:23:31.728379Z	Cloning repository...
2024-02-20T20:23:32.61015Z	From https://github.com/morganfeeney/minimal-repro
2024-02-20T20:23:32.610692Z	 * branch            24f7c351a3aa5c830204beecc0acd750c202b111 -> FETCH_HEAD
2024-02-20T20:23:32.610837Z	
2024-02-20T20:23:32.649766Z	HEAD is now at 24f7c35 Add og route
2024-02-20T20:23:32.650217Z	
2024-02-20T20:23:32.750971Z	
2024-02-20T20:23:32.751492Z	Using v2 root directory strategy
2024-02-20T20:23:32.780861Z	Success: Finished cloning repository files
2024-02-20T20:23:33.50549Z	Detected the following tools from environment: pnpm@8.7.1, nodejs@18.17.1
2024-02-20T20:23:34.416901Z	Installing project dependencies: pnpm install
2024-02-20T20:23:35.163643Z	Lockfile is up to date, resolution step is skipped
2024-02-20T20:23:35.181726Z	Progress: resolved 1, reused 0, downloaded 0, added 0
2024-02-20T20:23:35.218656Z	Packages: +139
2024-02-20T20:23:35.218983Z	++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
2024-02-20T20:23:36.195989Z	Progress: resolved 139, reused 0, downloaded 42, added 35
2024-02-20T20:23:37.200218Z	Progress: resolved 139, reused 0, downloaded 135, added 135
2024-02-20T20:23:38.200754Z	Progress: resolved 139, reused 0, downloaded 136, added 136
2024-02-20T20:23:39.201935Z	Progress: resolved 139, reused 0, downloaded 139, added 138
2024-02-20T20:23:39.218592Z	Progress: resolved 139, reused 0, downloaded 139, added 139, done
2024-02-20T20:23:39.350293Z	
2024-02-20T20:23:39.350918Z	dependencies:
2024-02-20T20:23:39.351147Z	+ next 14.1.0
2024-02-20T20:23:39.351372Z	+ react 18.2.0
2024-02-20T20:23:39.351589Z	+ react-dom 18.2.0
2024-02-20T20:23:39.351744Z	
2024-02-20T20:23:39.35191Z	devDependencies:
2024-02-20T20:23:39.352065Z	+ @types/node 20.11.19
2024-02-20T20:23:39.352245Z	+ @types/react 18.2.57
2024-02-20T20:23:39.352367Z	+ @types/react-dom 18.2.19
2024-02-20T20:23:39.352474Z	+ autoprefixer 10.4.17
2024-02-20T20:23:39.352663Z	+ postcss 8.4.35
2024-02-20T20:23:39.352809Z	+ tailwindcss 3.4.1
2024-02-20T20:23:39.352957Z	+ typescript 5.3.3
2024-02-20T20:23:39.353058Z	
2024-02-20T20:23:39.362813Z	Done in 4.5s
2024-02-20T20:23:39.521466Z	Executing user command: npx @cloudflare/next-on-pages@1
2024-02-20T20:23:41.86799Z	npm WARN exec The following package was not found and will be installed: @cloudflare/next-on-pages@1.9.0
2024-02-20T20:23:47.953024Z	npm WARN deprecated rollup-plugin-inject@3.0.2: This package has been deprecated and is no longer maintained. Please use @rollup/plugin-inject.
2024-02-20T20:23:48.042189Z	npm WARN deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
2024-02-20T20:23:49.331959Z	npm WARN deprecated uuid@3.3.2: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
2024-02-20T20:23:49.429153Z	npm WARN deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
2024-02-20T20:23:52.371689Z	⚡️ @cloudflare/next-on-pages CLI v.1.9.0
2024-02-20T20:23:52.756588Z	⚡️ Detected Package Manager: npm (9.6.7)
2024-02-20T20:23:52.75725Z	⚡️ Preparing project...
2024-02-20T20:23:52.757618Z	
2024-02-20T20:23:52.757952Z	⚡️ The project is set up for pnpm but it is currently being run via npm this might lead to build errors, please be sure to use the same package manager your project uses when running @cloudflare/next-on-pages
2024-02-20T20:23:52.758216Z	
2024-02-20T20:23:52.760955Z	⚡️ Project is ready
2024-02-20T20:23:52.761175Z	⚡️ Building project...
2024-02-20T20:23:53.909421Z	▲  npm WARN exec
2024-02-20T20:23:53.909901Z	▲  The following package was not found and will be installed: vercel@33.5.1
2024-02-20T20:23:56.167395Z	▲  npm WARN
2024-02-20T20:23:56.168225Z	▲  deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
2024-02-20T20:23:56.222385Z	▲  npm WARN
2024-02-20T20:23:56.222722Z	▲  deprecated uuid@3.3.2: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
2024-02-20T20:23:57.623038Z	▲  Vercel CLI 33.5.1
2024-02-20T20:23:57.927617Z	▲  Detected `pnpm-lock.yaml` version 6 generated by pnpm 8
2024-02-20T20:23:57.940248Z	▲  Installing dependencies...
2024-02-20T20:23:58.484046Z	▲  Lockfile is up to date, resolution step is skipped
2024-02-20T20:23:58.513541Z	▲  Already up to date
2024-02-20T20:23:58.643382Z	▲  
2024-02-20T20:23:58.654102Z	▲  Done in 571ms
2024-02-20T20:23:58.673608Z	▲  Detected Next.js version: 14.1.0
2024-02-20T20:23:58.678444Z	▲  Running "pnpm run build"
2024-02-20T20:23:59.149838Z	▲  > minimal@0.1.0 build /opt/buildhome/repo
2024-02-20T20:23:59.150106Z	▲  > next build
2024-02-20T20:23:59.871157Z	▲  Attention: Next.js now collects completely anonymous telemetry regarding usage.
2024-02-20T20:23:59.871769Z	▲  This information is used to shape Next.js' roadmap and prioritize features.
2024-02-20T20:23:59.871941Z	▲  You can learn more, including how to opt-out if you'd not like to participate in this anonymous program, by visiting the following URL:
2024-02-20T20:23:59.872122Z	▲  https://nextjs.org/telemetry
2024-02-20T20:23:59.872234Z	▲  
2024-02-20T20:23:59.986483Z	▲  ▲ Next.js 14.1.0
2024-02-20T20:23:59.986778Z	▲  
2024-02-20T20:24:00.047776Z	▲  Creating an optimized production build ...
2024-02-20T20:24:15.127131Z	▲  ✓ Compiled successfully
2024-02-20T20:24:15.128323Z	▲  Linting and checking validity of types ...
2024-02-20T20:24:17.311871Z	▲  Collecting page data ...
2024-02-20T20:24:18.359251Z	▲  ⚠ Using edge runtime on a page currently disables static generation for that page
2024-02-20T20:24:18.815966Z	▲  Generating static pages (0/5) ...
2024-02-20T20:24:19.515324Z	▲  Generating static pages (1/5)
2024-02-20T20:24:19.79326Z	▲  Generating static pages (2/5)
2024-02-20T20:24:19.793618Z	▲  Generating static pages (3/5)
2024-02-20T20:24:19.819428Z	▲  ✓ Generating static pages (5/5)
2024-02-20T20:24:20.182784Z	▲  Finalizing page optimization ...
2024-02-20T20:24:20.183057Z	▲  Collecting build traces ...
2024-02-20T20:24:26.923396Z	▲  
2024-02-20T20:24:26.926926Z	▲  Route (app)                              Size     First Load JS
2024-02-20T20:24:26.92717Z	▲  ┌ ○ /                                    5.13 kB        89.3 kB
2024-02-20T20:24:26.927342Z	▲  ├ ○ /_not-found                          885 B          85.1 kB
2024-02-20T20:24:26.927501Z	▲  └ ℇ /og                                  0 B                0 B
2024-02-20T20:24:26.927649Z	▲  + First Load JS shared by all            84.2 kB
2024-02-20T20:24:26.927765Z	▲  ├ chunks/184-d3bb186aac44da98.js       28.9 kB
2024-02-20T20:24:26.927902Z	▲  ├ chunks/30b509c0-f3503c24f98f3936.js  53.4 kB
2024-02-20T20:24:26.928049Z	▲  └ other shared chunks (total)          1.87 kB
2024-02-20T20:24:26.928285Z	▲  
2024-02-20T20:24:26.928536Z	▲  ○  (Static)        prerendered as static content
2024-02-20T20:24:26.928679Z	▲  ℇ  (Edge Runtime)  server-rendered on demand using the Edge Runtime
2024-02-20T20:24:27.579095Z	▲  Traced Next.js server files in: 570.277ms
2024-02-20T20:24:28.593574Z	▲  Created all serverless functions in: 1.014s
2024-02-20T20:24:28.830074Z	▲  Collected static files (public/, static/, .next/static): 11.252ms
2024-02-20T20:24:28.890061Z	▲  Build Completed in .vercel/output [31s]
2024-02-20T20:24:28.991362Z	⚡️ Completed `npx vercel build`.
2024-02-20T20:24:29.361632Z	⚡️ Unexpected error: ENOENT: no such file or directory, copyfile '/opt/buildhome/repo/.vercel/output/functions/og.func/wasm/wasm_77d9faebf7af9e421806970ce10a58e9d83116d7.wasm' -> '/opt/buildhome/repo/.vercel/output/static/_worker.js/__next-on-pages-dist__/wasm/wasm_77d9faebf7af9e421806970ce10a58e9d83116d7.wasm'
2024-02-20T20:24:29.425953Z	Finished
2024-02-20T20:24:29.426534Z	Found _worker.js in output directory. Uploading.
2024-02-20T20:24:30.813567Z	[31m✘ [41;31m[[41;97mERROR[41;31m][0m [1m1 error(s) and 0 warning(s) when compiling Worker.[0m
2024-02-20T20:24:30.814248Z	
2024-02-20T20:24:30.814442Z	
2024-02-20T20:24:30.816971Z	
2024-02-20T20:24:30.818231Z	[31m✘ [41;31m[[41;97mERROR[41;31m][0m [1mCould not resolve "/opt/buildhome/repo/.vercel/output/static/_worker.js/index.js"[0m
2024-02-20T20:24:30.818414Z	
2024-02-20T20:24:30.818677Z	
2024-02-20T20:24:30.820161Z	[31m✘ [41;31m[[41;97mERROR[41;31m][0m [1mBuild failed with 1 error:[0m
2024-02-20T20:24:30.820663Z	
2024-02-20T20:24:30.820853Z	  error: Could not resolve "/opt/buildhome/repo/.vercel/output/static/_worker.js/index.js"
2024-02-20T20:24:30.820998Z	
2024-02-20T20:24:30.821117Z	
2024-02-20T20:24:30.833329Z	Failed building Pages Functions.
2024-02-20T20:24:31.825753Z	Failed: generating Pages Functions failed. Check the logs above for more information. If this continues for an unknown reason, contact support: https://cfl.re/3WgEyrH

### Would you like to help?

- [ ] Would you like to help fixing this bug?

### LINKED ISSUE #676: [🐛 Bug]: wasm imports don't work
### next-on-pages environment related information

```
System:
        Platform: linux
        Arch: x64
        Version: #1 SMP Thu Oct 5 21:02:42 UTC 2023
        CPU: (12) x64 AMD Ryzen 5 5500U with Radeon Graphics
        Memory: 3 GB
        Shell: /bin/bash
Package Manager Used: npm (10.4.0)

Relevant Packages:
        @cloudflare/next-on-pages: 1.9.0
        vercel: 33.4.1
        next: 14.0.4
```

### Description

The build using `npx @cloudflare/next-on-pages` failing when wasm modules are imported within the app.
I have imported modules (`wasm/photon/index.ts` and `wasm/png/index.ts`) in `app/api/image/route.ts` which imports wasm internally.
Everything works as expected in local development using `next dev`.
But when we try to build it for cloudflare pages using `npx @cloudflare/next-on-pages`, it shows the following error:
```
⚡️ @cloudflare/next-on-pages CLI v.1.9.0
⚡️ Detected Package Manager: npm (10.4.0)
⚡️ Preparing project...
⚡️ Project is ready
⚡️ Building project...
▲  Vercel CLI 33.4.1
▲  WARNING: You should not upload the `.next` directory.
▲  Installing dependencies...
▲  up to date in 587ms
▲  28 packages are looking for funding
▲  run `npm fund` for details
▲  Detected Next.js version: 14.0.4
▲  Detected `package-lock.json` generated by npm 7+
▲  Running "npm run build"
▲  > next-on-pages-cloudflare-global-example@0.1.0 build
▲  > next build
▲  ▲ Next.js 14.0.4
▲  Creating an optimized production build ...
▲  ✓ Compiled successfully
▲  Linting and checking validity of types ...
▲  Collecting page data ...
▲  ⚠ Using edge runtime on a page currently disables static generation for that page
▲  Generating static pages (0/4) ...
▲  Generating static pages (1/4)
▲  Generating static pages (2/4)
▲  Generating static pages (3/4)
▲  ✓ Generating static pages (4/4)
▲  Finalizing page optimization ...
▲  Collecting build traces ...
▲  
▲  Route (app)                              Size     First Load JS
▲  ┌ ○ /                                    335 B          82.2 kB
▲  ├ ℇ /_not-found                          0 B                0 B
▲  ├ ℇ /api/hello                           0 B                0 B
▲  ├ ℇ /api/image                           0 B                0 B
▲  ├ ℇ /client                              620 B          82.5 kB
▲  └ ℇ /server                              141 B            82 kB
▲  + First Load JS shared by all            81.9 kB
▲  ├ chunks/993-25f0712dd7de98d2.js       26.7 kB
▲  ├ chunks/fd9d1056-735d320b4b8745cb.js  53.3 kB
▲  ├ chunks/main-app-f0368426e1af5d32.js  242 B
▲  └ chunks/webpack-3b5a3d1c5cd9ecf8.js   1.69 kB
▲  ○  (Static)        prerendered as static content
▲  ℇ  (Edge Runtime)  server-rendered on demand using the Edge Runtime
▲  Traced Next.js server files in: 501.244ms
▲  Created all serverless functions in: 858.896ms
▲  Collected static files (public/, static/, .next/static): 5.552ms
▲  Build Completed in .vercel/output [21s]
⚡️ Completed `npx vercel build`.
⚡️ Unexpected error: ENOENT: no such file or directory, copyfile '/home/deo/Pages/next-on-pages-get-request-context-test-demo/.vercel/output/functions/api/image.func/wasm/wasm_65d5384f60effcb9a86d814043fa7bb1fdd99f0b.wasm' -> '/home/deo/Pages/next-on-pages-get-request-context-test-demo/.vercel/output/static/_worker.js/__next-on-pages-dist__/wasm/wasm_65d5384f60effcb9a86d814043fa7bb1fdd99f0b.wasm'
```

### Reproduction

1. Clone the repo:
   ```
   git clone https://github.com/techwithdeo/next-cf-pages-wasm.git
   ```
2. Choose the directory:
   ```
   cd next-cf-pages-wasm
   ```
3. Install dependencies
   ```
   npm install
   ```
4. Start local development:
   ```
   npm run dev
   ```
5. Open the path in browser (works as expected):
   ```
   /api/image
   ```
6. Now build it using `@cloudflare/next-on-pages` (shows error):
   ```
   npm run pages:build
   ```

### Pages Deployment Method

None

### Pages Deployment ID

Not yet deployed

### Additional Information

Note that the build is successful when routes which import wasm are removed/deleted.

### Would you like to help?

- [ ] Would you like to help fixing this bug?

### PR DESCRIPTION
**Pull request by james-elicx**
This PR does the following:
- Introduces some support for the `filePathMap` option in a function config file.
- Symlinks assets to the function's folder, so that they can be found during the build process.

fixes #676
fixes #672

Example of this in a function's `.vc-config.json`:

image

### CODE REVIEW

**Comment by karthikjn01** [comment]
I've given this a test on my system, and it works! 🙌

**Comment by fineshop** [comment]
Thanks, wasm imports work as expected now.

File: packages/next-on-pages/src/buildApplication/processVercelFunctions/edgeFunctions.ts [hunk 1]
```diff
@@ -75,6 +76,38 @@ export async function processEdgeFunctions(
...
+			const currentAssetPath = resolve(assetLocation);
+			const newAssetPath = join(path, relativeAssetPath);
+
+			if (!(await validateFile(currentAssetPath))) {
+				cliError(`Could not find bundled asset file: ${assetLocation}`);
+				process.exit(1);
+			}
+
+			await mkdir(dirname(newAssetPath), { recursive: true });
+			await symlink(currentAssetPath, newAssetPath, 'file');
```
Comment (dario-piotrowicz) [inline packages/next-on-pages/src/buildApplication/processVercelFunctions/edgeFunctions.ts:108]: I think this breaks when using the `-s|--skip-build` flag:
Screenshot 2024-02-27 at 23 40 45

**Review by dario-piotrowicz — COMMENTED** [review]
Thanks a lot @james-elicx

I've seen this issue popup here and there and hadn't had the time to tackle it, thanks so much for having a look 🙏 

But I am a bit hesitant on the changes for two reasons, one is the `--skip-build` flag issue I mentioned in my comment (should be an easy fix anyways) the other is that I am not sure if changing any file outside `.vercel/static/output` is appropriate... 

I might be wrong but I think we don't do this anywhere else and try to keep things in memory instead whenever possible (to avoid issues like the one I mentioned in which using the skip build option can make the build no longer idempotent)

what do you think?

File: packages/next-on-pages/src/buildApplication/processVercelFunctions/edgeFunctions.ts (hunk 1 above)
Comment (james-elicx) [reply packages/next-on-pages/src/buildApplication/processVercelFunctions/edgeFunctions.ts:108]: ah good catch, adding a check to only do it when there is no file there.

**Comment by james-elicx** [comment]
> the other is that I am not sure if changing any file outside .vercel/static/output is appropriate...
>
> I might be wrong but I think we don't do this anywhere else and try to keep things in memory instead whenever possible (to avoid issues like the one I mentioned in which using the skip build option can make the build no longer idempotent)
>
> what do you think?

this is true. i dont think we do this anywhere else in the .vercel/output/functions. on the one hand, this make the fix a lot simpler, instead of changing how some of the deeper logic works, as it allows all the logic we have for the assets in a function's dir to continue doing their thing instead.

would you prefer i make the change to the deeper parts where it's processed instead of symlinking them?

**Comment by dario-piotrowicz** [comment]
> would you prefer i make the change to the deeper parts where it's processed instead of symlinking them?

yeah I think I'd much rather prefer that if it's not too much of a thing to ask 😓 

I feel like if we start to "cheat" things can go out of hand and spiral out of control, saying "we never touch files that don't belong to us" feels like a much cleaner/clearer mental model that can prevent pain in the long run (and again, I think that adding exceptions to this can lead to unpredicability and pain 😓)

I don't think that not using symlinks is going to be problematic... it might a bit cumbersome as the paths need to be passed around though 😓 

Would you be willing to give that a shot? 🙏 
(if you prefer you can also do it in a separate branch and PR and we can compare that solution to this one)

**Comment by james-elicx** [comment]
> Would you be willing to give that a shot? 🙏

There ya go. You were right, should have done it like this originally, good shout :)

**Review by dario-piotrowicz — APPROVED** [review]
Thanks so much, yeah now it looks perfect 😍



> Sorry for the delay!

no problem at all 😄, thanks for the fix! 🙏

**Comment by fineshop** [comment]
Thank you all for the fix! 🙂

**Pull request merged**